<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Tom Michoel">
<meta name="dcterms.date" content="2023-09-14">

<title>BINF200 Assignment 1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="assignment_PacBio_VSG_files/libs/clipboard/clipboard.min.js"></script>
<script src="assignment_PacBio_VSG_files/libs/quarto-html/quarto.js"></script>
<script src="assignment_PacBio_VSG_files/libs/quarto-html/popper.min.js"></script>
<script src="assignment_PacBio_VSG_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="assignment_PacBio_VSG_files/libs/quarto-html/anchor.min.js"></script>
<link href="assignment_PacBio_VSG_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="assignment_PacBio_VSG_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="assignment_PacBio_VSG_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="assignment_PacBio_VSG_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="assignment_PacBio_VSG_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">BINF200 Assignment 1</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Tom Michoel </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Due date</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 14, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="grading-and-report" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="grading-and-report"><span class="header-section-number">1</span> Grading and report</h2>
<p>The assignment is scored on 10 points and counts towards 10% of the final grade.</p>
</section>
<section id="background" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="background"><span class="header-section-number">2</span> Background</h2>
<p>In this assignment you will analyze data from the following paper:</p>
<p>S. Jayaraman <em>et al.</em> (2019), <a href="https://doi.org/10.1371/journal.pntd.0007262">Application of long read sequencing to determine expressed antigen diversity in Trypanosoma brucei infections</a>, PLOS Neglected Tropical Diseases 13(4): e0007262.</p>
<p>You will need a copy of the paper to answer some of the assignment question.</p>
<p>The paper has an accompanying <a href="https://github.com/siddharthjayaraman/longread-application">Github repository</a> where you will find some useful commands.</p>
</section>
<section id="software" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="software"><span class="header-section-number">3</span> Software</h2>
<p>You will need the following software to solve the assignment:</p>
<ol type="1">
<li><p><strong>BLAST+</strong>: Install the NCBI command line standalone BLAST+ programs following the <a href="https://www.ncbi.nlm.nih.gov/books/NBK52640/">online instructions</a>.</p>
<ul>
<li>On Windows, installation in the <a href="https://learn.microsoft.com/en-us/windows/wsl/">Windows Subsystem for Linux</a> will be easier than following the above instructions.</li>
<li>For Ubuntu (and hence also WSL), there exists a package “ncbi-blast+” not listed on the NCBI page, which you can install directly using <code>apt-get install</code>.</li>
</ul></li>
<li><p><strong>GetORF</strong>: Install the <a href="https://bio.tools/getorf">getorf</a> tool. The <a href="http://emboss.open-bio.org/html/adm/ch01s01.html">installation instructions</a> are a bit tedious, but you should get there in the end. Again, on Windows, installation in the WSL will be easier.</p></li>
<li><p><strong>Programming language</strong>: You may use any language: Python, Julia, R, …</p></li>
<li><p><strong>Packages</strong> for reading FASTA-formatted files and importing and manipulating DataFrames. Find and install them for your preferred programming language.</p></li>
<li><p><strong>Data analysis and visualization, report writing</strong>: Use <a href="https://jupyter.org/">JupyterLab</a> (or Jupyter notebooks) and/or an IDE such as <a href="https://code.visualstudio.com/">Visual Studio Code</a> for data analysis and visualization. Using <a href="https://quarto.org/">Quarto</a> you can integrate your notebooks, data visualization, and final report all in a markdown document. <a href="https://plutojl.org/">Pluto</a> notebooks for Julia offer nice reactive features.</p></li>
</ol>
</section>
<section id="sequence-data" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="sequence-data"><span class="header-section-number">4</span> Sequence data</h2>
<p>All data for the assignment are available on <a href="https://universityofbergen-my.sharepoint.com/:f:/g/personal/tom_michoel_uib_no/EphdekZYkYBJnqYKRcpoOfsBJBouvZhTOQEorETwKBHGsA?e=roErlx">UiB OneDrive</a> in the <strong>PacBio VSG</strong> folder. You will need to login with your UiB account to obtain access. You will need the following files:</p>
<ol type="1">
<li><p>A database of known VSG genes. Download the entire folder <strong>TREU927-v26_VSGTranscripts</strong>.</p></li>
<li><p>Sequencing data from <strong>one</strong> sample (<strong>one</strong> file) named <strong>PacBio_VSG_filtered_reads_<em>sample_name</em>.fasta</strong>. Twenty such files, for 20 samples (individual mice), are available in total. To find out which one <em>you</em> should download:</p>
<ol type="1">
<li>Go to the <strong>Groups</strong> page on Mitt (<a href="https://mitt.uib.no/courses/42444/groups">https://mitt.uib.no/courses/42444/groups</a>)</li>
<li>Select the <strong>Compulsory Assignment 1 - Sample</strong> tab and find to which group you have been (randomly) assigned.</li>
<li>Each group is named <strong>Compulsory Assignment 1 - Sample <em>k</em></strong>, where <em>k</em> is a number from 1 to 20 mapping to the sample names as follows:</li>
</ol></li>
</ol>
<div id="tbl-groups" class="anchored">
<table class="table">
<caption>Table&nbsp;1: Mapping of group labels to sample names</caption>
<thead>
<tr class="header">
<th>k</th>
<th>Sample name</th>
<th>k</th>
<th>Sample name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>balbc_3_0</td>
<td>11</td>
<td>balbc_10_0</td>
</tr>
<tr class="even">
<td>2</td>
<td>balbc_3_1</td>
<td>12</td>
<td>balbc_10_1</td>
</tr>
<tr class="odd">
<td>3</td>
<td>balbc_3_2</td>
<td>13</td>
<td>balbc_10_2</td>
</tr>
<tr class="even">
<td>4</td>
<td>balbc_3_3</td>
<td>14</td>
<td>balbc_10_4</td>
</tr>
<tr class="odd">
<td>5</td>
<td>balbc_3_4</td>
<td>15</td>
<td>balbc_10_5</td>
</tr>
<tr class="even">
<td>6</td>
<td>balbc_6_0</td>
<td>16</td>
<td>balbc_12_1</td>
</tr>
<tr class="odd">
<td>7</td>
<td>balbc_6_1</td>
<td>17</td>
<td>balbc_12_2</td>
</tr>
<tr class="even">
<td>8</td>
<td>balbc_6_2</td>
<td>18</td>
<td>balbc_12_3</td>
</tr>
<tr class="odd">
<td>9</td>
<td>balbc_6_4</td>
<td>19</td>
<td>balbc_12_4</td>
</tr>
<tr class="even">
<td>10</td>
<td>balbc_6_5</td>
<td>20</td>
<td>balbc_12_5</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="tasks" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="tasks"><span class="header-section-number">5</span> Tasks</h2>
<section id="show-that-you-understand-the-blast-package" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="show-that-you-understand-the-blast-package"><span class="header-section-number">5.1</span> Show that you understand the BLAST+ package</h3>
<p>The BLAST+ package contains five <strong>core blast search programs</strong>.</p>
<ol type="1">
<li>List all five core blast search programs.</li>
<li>Explain the difference between <strong>blastn</strong> and <strong>blastp</strong>.</li>
</ol>
</section>
<section id="show-that-you-understand-the-biological-experiment-and-data" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="show-that-you-understand-the-biological-experiment-and-data"><span class="header-section-number">5.2</span> Show that you understand the biological experiment and data</h3>
<p>Write a short paragraph in your own words to explain the biological experiment that was done to generate the sequence data (see the <a href="https://doi.org/10.1371/journal.pntd.0007262">publication</a>), what sequences are contained in the <strong>PacBio_VSG_filtered_reads_<em>sample_name</em>.fasta</strong> files, and what sequences are contained in the <strong>TREU927-v26_VSGTranscripts</strong> database.</p>
<p>Answer the following question: What is the number of sequences in <em>your</em> input file?</p>
<p><strong>Hint:</strong> Most popular programming languages have packages that can parse FASTA files automatically.</p>
</section>
<section id="blast-the-sample-sequences-against-the-reference-vsg-database" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="blast-the-sample-sequences-against-the-reference-vsg-database"><span class="header-section-number">5.3</span> BLAST the sample sequences against the reference VSG database</h3>
<p>Find the relevant BLAST command on the <a href="https://github.com/siddharthjayaraman/longread-application">longread-application repository</a>. Then adapt this command in the following ways:</p>
<ol type="1">
<li><p>Change the input file name to <em>your</em> assigned sample file (see <a href="#tbl-groups">Table&nbsp;1</a>).</p></li>
<li><p>Change the output file name to something containing <em>your</em> sample label (see <a href="#tbl-groups">Table&nbsp;1</a>).</p></li>
<li><p>What does the paramater <code>-max_target_seq</code> do and why was it set to 1?</p></li>
<li><p>Use a tabular output format <em>without</em> comments that includes <em>only</em> the following columns:</p>
<ul>
<li>Query sequence id</li>
<li>Subject sequence id</li>
<li>Raw score</li>
<li>Bit score</li>
<li>E-value</li>
<li>Query sequence length</li>
<li>Subject sequence length</li>
<li>Alignment length</li>
<li>Start of alignment in subject</li>
<li>End of alignment in subject</li>
<li>Number of identical matches</li>
<li>Number of mismatches</li>
<li>Total number of gaps</li>
<li>Percentage of positive-scoring matches</li>
</ul></li>
</ol>
<p>Include the exact BLAST command you ran in your report.</p>
</section>
<section id="analyze-the-blast-output" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="analyze-the-blast-output"><span class="header-section-number">5.4</span> Analyze the BLAST output</h3>
<p>Answer the following questions / perform the following tasks:</p>
<ol type="1">
<li><p>What is the number of sequences in <em>your</em> blast output file? Is it the same as in the input file?<br> <strong>Hint:</strong> Import the blastn output file into a DataFrame.</p></li>
<li><p>We define the alignment coverage as the percentage of the subject sequence covered by the alignment. Compute the alignment coverage for all sequences from the blastn output.</p></li>
<li><p>Visualize the distribution of alignment coverages as a histogram.</p></li>
<li><p>Remove alignments with coverage less than 60% and verify that each query sequence is now aligned to at most one subject sequence.</p></li>
<li><p>The bit score <span class="math inline">\(S'\)</span> is derived from the raw score <span class="math inline">\(S\)</span> using the formula <span class="math display">\[
   S' = \frac{\lambda S - \ln K}{\ln 2}
   \]</span> Can you find the values of <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(K\)</span> from your blast results?</p></li>
</ol>
</section>
<section id="count-vsg-expression-levels" class="level3" data-number="5.5">
<h3 data-number="5.5" class="anchored" data-anchor-id="count-vsg-expression-levels"><span class="header-section-number">5.5</span> Count VSG expression levels</h3>
<p>Perform the following tasks:</p>
<ol type="1">
<li><p>Extract the unique VSG ids in <em>your</em> sample from your (filtered) BLAST results.</p></li>
<li><p>For each unique VSG: count the number of sequences aliging to that VSG, and the average number of identical matches, mismatches, and gaps for its alignments.<br> <strong>Hint:</strong> Using the split-apply-combine strategy, these numbers can be computed in one line of code.</p></li>
<li><p>Identify the 10 most abundant VSGs in <em>your</em> sample and visualize their relative expression levels in a pie chart and compare your result against <a href="https://doi.org/10.1371/journal.pntd.0007262.g003">Figure 3</a> of the <a href="https://doi.org/10.1371/journal.pntd.0007262">paper</a>.</p></li>
<li><p>Write a paragraph in your report that describes the figures and your interpretation of them.</p></li>
</ol>
</section>
<section id="identify-open-reading-frames" class="level3" data-number="5.6">
<h3 data-number="5.6" class="anchored" data-anchor-id="identify-open-reading-frames"><span class="header-section-number">5.6</span> Identify open reading frames</h3>
<p>Perform the following tasks:</p>
<ol type="1">
<li><p>Find the relevant <code>getorf</code> command on the <a href="https://github.com/siddharthjayaraman/longread-application">longread-application repository</a> and adapt it to work on <em>your</em> input sample.</p></li>
<li><p>Count the percentage of reads in <em>your</em> sample that result in a predicted ORF with a minimum size of 1200 nucleotides.</p></li>
<li><p>Which explanation was proposed in the <a href="https://doi.org/10.1371/journal.pntd.0007262">paper</a> for this low percentage?</p></li>
</ol>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
</div> <!-- /content -->



</body></html>